Updated README for Label dashboard profile sept 2023 (#1037)
e-mission/e-mission-phone: README.md

# [e-mission phone app](https://github.com/e-mission/e-mission-phone/tree/master)

__This is the phone component of the e-mission system.__

> :sparkles: This has now been upgraded to cordova android@12.0.0 and iOS@6.2.0. It has also been upgraded to the **latest Android & iOS versions**, **cordova-lib@10.0.0 and the most recent node and npm versions**. It also now supports CI, so we should not have any build issues in the future. __This should be ready to build out of the box.__

Reviewer: Why have the versions not been removed?
https://github.com/e-mission/e-mission-phone/pull/1037#discussion_r1335066508

✨ We constantly upgrade the repo to the latest cordova versions of android, iOS, cordova-lib, and the most recent node and npm versions. The CI will be up-to-date. 

> For the latest versions, refer [`package.cordovabuild.json`](https://github.com/e-mission/e-mission-phone/blob/fce117ff859abd995613bd405dbc7d27c703b09b/package.cordovabuild.json)

Reviewer: having this include a permalink defeats the purpose of using it to check the latest links

Author: Ah shoot. Would relative path be a good substitute?

## Additional Documentation
Additional documentation has been moved to its own repository [e-mission-docs](https://github.com/e-mission/e-mission-docs). Specific e-mission-phone wikis can be found here:
https://github.com/e-mission/e-mission-docs/tree/master/docs/e-mission-phone

**Issues:** Since this repository is part of a larger project, all issues are tracked [in the central docs repository](https://github.com/e-mission/e-mission-docs/issues). If you have a question, [as suggested by the open source guide](https://opensource.guide/how-to-contribute/#communicating-effectively), please file an issue instead of sending an email. Since issues are public, other contributors can try to answer the question and benefit from the answer.

:sparkles: Check [Contributing](#contributing) if you're interested in contributing for this project :sparkles:

## Contents
#### 1. [Updating the UI only](#updating-the-ui-only)
#### 2. [Updating the e-mission-* plugins or adding new plugins](#updating-the-e-mission--plugins-or-adding-new-plugins)  

> #### 3. [Creating logos](#creating-logos) 

Reviewer: why is creating logos above end to end testing? Creating new logos is a very very unusual task

Reviewer: To add on to this, why is the ToC not consistent with the section order?

> #### 4. [End to End Testing](#end-to-end-testing)

Reviewer: why is end-to-end testing step 4? The common case so far is UI development. As I said before:
https://github.com/e-mission/e-mission-phone/pull/1037#discussion_r1335096417
it should be before the native build

#### 5. [Beta-testing debugging](#beta-testing-debugging) 
#### 6. [Contributing](#contributing)
#### 7. [Troubleshooting](#troubleshooting)

---

## Updating the UI only
[![osx-serve-install](https://github.com/e-mission/e-mission-phone/workflows/osx-serve-install/badge.svg)](https://github.com/e-mission/e-mission-phone/actions/workflows/serve-install.yml)

If you want to make only UI changes, (as opposed to modifying the existing plugins, adding new plugins, etc), you can use the **new and improved** (as of June 2018) [e-mission dev app](https://github.com/e-mission/e-mission-devapp/) and install the most recent version from [releases](https://github.com/e-mission/e-mission-devapp/releases). 

### Installing (one-time)

1. Run the setup script

```
bash setup/setup_serve.sh
```
**(optional)** Configure by changing the files in `www/json`.
Defaults are in `www/json/*.sample`

```
ls www/json/*.sample
cp www/json/startupConfig.json.sample www/json/startupConfig.json

> cp ..... www/json/connectionConfig.json

Reviewer: I have not received any links from either you or Jack. Please revert all these changes; we can include them in a future README change.

Author: https://github.com/e-mission/e-mission-phone/pull/1037/files/bbf18bdb939f9a30670cda10c54de4c66a251c84#r1342002279

# [e-mission phone app](https://github.com/e-mission/e-mission-phone/tree/master)

<u>__This is the phone component of the e-mission system.__</u>

> :sparkles: This has now been upgraded to cordova android@12.0.0 and iOS@6.2.0. It has also been upgraded to [android API 33 and the latest iOS versions](https://github.com/e-mission/e-mission-docs/issues/934), [cordova-lib@10.0.0 and the most recent node and npm versions](<insert link>). It also now supports CI, so we should not have any build issues in the future. __This should be ready to build out of the box.__

Reviewer 2: I'm on board with not listing the version numbers and just telling people to look at `package.cordovabuild.json` and `package.serve.json`.

Overall I think this paragraph is 'fluff' that gets in the way and can be removed

#### 4. [End to End Testing](#4-end-to-end-testing)
#### 5. [Beta-testing debugging](#5-beta-testing-debugging) 
#### 6. [Contributing](#6-contributing)

---

## 1. Creating logos

If you are building your own version of the app, you must have your own logo to
avoid app store conficts. Updating the logo is very simple using the [`ionic
cordova resources`](https://ionicframework.com/docs/v3/cli/cordova/resources/)
command.

**Note**: You may have to install the [`cordova-res` package](https://github.com/ionic-team/cordova-res) for the command to work

## 2. Updating the UI only
[![osx-serve-install](https://github.com/e-mission/e-mission-phone/workflows/osx-serve-install/badge.svg)](https://github.com/e-mission/e-mission-phone/actions?query=workflow%3Aosx-serve-install)

If you want to make only UI changes, (as opposed to modifying the existing plugins, adding new plugins, etc), you can use the **new and improved** (as of June 2018) [e-mission dev app](https://github.com/e-mission/e-mission-devapp/) and install the most recent version from [releases](https://github.com/e-mission/e-mission-devapp/releases). 

### Installing (one-time)

1. Run the setup script

```

> bash setup/setup_serve.sh

Reviewer 2: But then when you copy the command to your clipboard (with the copy button Github adds), the `$` gets included, annoyingly.
I regularly plenty of READMEs without `$` prefixing every command

> bash setup/setup_serve.sh

Reviewer: This seems like it is all the fault of the Github "copy" button 😄 
I am open to revisiting this but would like to see examples of the READMEs first.

> bash setup/setup_serve.sh

Author: Would these README examples work?
- https://github.com/airbnb/javascript
- https://github.com/NationalSecurityAgency/ghidra

I don't mind restoring it but like @JGreenlee said, it can be annoying to manually select the command everytime

> bash setup/setup_serve.sh

Reviewer: The first is irrelevant since it includes **code** samples, not command line samples.

The second is in fact relevant. If you can find a second example of CLI samples, I am happy to switch.

> bash setup/setup_serve.sh

Reviewer 2: 1. https://github.com/necolas/react-native-web/blob/master/.github/CONTRIBUTING.md
1. https://github.com/facebook/create-react-app/
1. https://github.com/enketo/enketo-transformer
1. TADA's certs repo

A lot of React / React Native docs are kept on their own website, not READMEs, but when they do show CLI samples they generally don't have `$` (e.g. https://callstack.github.io/react-native-paper/docs/guides/getting-started/)

> bash setup/setup_serve.sh

Reviewer: ok, I am convinced. @niccolopaganini You can keep remove the leading `$` in this PR.

```
**(optional)** Configure by changing the files in `www/json`.
Defaults are in `www/json/*.sample`

```
ls www/json/*.sample
cp www/json/startupConfig.json.sample www/json/startupConfig.json
cp ..... www/json/connectionConfig.json
```

### Activation (after install, and in every new shell)

> 2. Run this to activate 

Reviewer: This comment has not been addressed

> # [e-mission phone app](https://github.com/e-mission/e-mission-phone/tree/master)

Reviewer: why does this need to be a URL given that we are literally in the same repo?

__This is the phone component of the e-mission system.__

:sparkles: This has been upgraded to the latest **Android**, **iOS**, **cordova-lib**, **node** and **npm** versions. __This is ready to build out of the box.__

> ✨ We constantly upgrade the repo to the latest cordova versions of android, iOS, cordova-lib, and the most recent node and npm versions. The CI will be up-to-date. 

Reviewer: We don't *constantly* upgrade the repo. Not sure what you mean by "the CI will be up-to-date".

```suggestion
✨ We periodically upgrade the repo to the latest cordova versions of android, iOS, cordova-lib, and the most recent node and npm versions. The CI will be up-to-date. 
```

> For the latest versions, refer [`package.cordovabuild.json`](package.cordovabuild.json)

Reviewer: ```suggestion
The currently supported versions are in [`package.cordovabuild.json`](package.cordovabuild.json)
```

Reviewer 2: @niccolopaganini This has not been addressed

[![osx-serve-install](https://github.com/e-mission/e-mission-phone/workflows/osx-serve-install/badge.svg)](https://github.com/e-mission/e-mission-phone/actions/workflows/serve-install.yml)

If you want to make only UI changes, (as opposed to modifying the existing plugins, adding new plugins, etc), you can use the **new and improved** (as of June 2018) [e-mission dev app](https://github.com/e-mission/e-mission-devapp/) and install the most recent version from [releases](https://github.com/e-mission/e-mission-devapp/releases). 

### Installing (one-time)

Run the setup script
:point_right:Run the setup script

```
$ bash setup/setup_serve.sh
bash setup/setup_serve.sh
```

**(optional)** Configure by changing the files in `www/json`.
Defaults are in `www/json/*.sample`

```
$ ls www/json/*.sample
$ cp www/json/startupConfig.json.sample www/json/startupConfig.json
$ cp ..... www/json/connectionConfig.json
ls www/json/*.sample
cp www/json/startupConfig.json.sample www/json/startupConfig.json
cp ..... www/json/connectionConfig.json
```

### Activation (after install, and in every new shell)

```
$ source setup/activate_serve.sh
source setup/activate_serve.sh
```
  
### Running

> 1. Start the phonegap deployment server and note the URL(s) that the server is listening to.

Reviewer: In this case, we had numbering. Please restore. Your changes should be restricted to the ones here:
https://github.com/e-mission/e-mission-phone/pull/1037#pullrequestreview-1651615631

Start the phonegap deployment server and note the URL(s) that the server is listening to.

>     ```

Reviewer: why has the code block been removed?

**Note1**: You may need to scroll up, past all the warnings about `Content Security Policy has been added` to find the port that the server is listening to.

## End to End Testing

> 

Reviewer: Extraneous change

:sparkles: Check [Contributing](#6-contributing) if you're interested in contributing for this project :sparkles:

## Contents
#### 1. [Creating logos](#creating-logos) 
#### 2. [Updating the UI only](#updating-the-ui-only) 
#### 3. [Updating the e-mission-* plugins or adding new plugins](#updating-the-e-mission--plugins-or-adding-new-plugins) 
#### 4. [End to End Testing](#end-to-end-testing)
#### 5. [Beta-testing debugging](#beta-testing-debugging) 

> #### 6. [Contributing](#contributing)

Reviewer: I don't see the benefit of removing the numbers but still keeping this order.

    ....
    ```
  
2. Change the devapp connection URL to one of these (e.g. 192.168.162.1:3000) and press "Connect"

> 3. The app will now display the version of e-mission app that is in your local directory

Reviewer: This comment has not been addressed

**Note1**: You may need to scroll up, past all the warnings about `Content Security Policy has been added` to find the port that the server is listening to.

## 3. Updating the e-mission-\* plugins or adding new plugins

[![osx-build-ios](https://github.com/e-mission/e-mission-phone/actions/workflows/ios-build.yml/badge.svg)](https://github.com/e-mission/e-mission-phone/actions/workflows/ios-build.yml)
[![osx-build-android](https://github.com/e-mission/e-mission-phone/actions/workflows/android-build.yml/badge.svg)](https://github.com/e-mission/e-mission-phone/actions/workflows/android-build.yml)

__Important__

Most of the recent issues encountered have been due to incompatible setup. We
have now:
- locked down the dependencies,
- created setup and teardown scripts to setup self-contained environments with
  those dependencies, and
- CI enabled to validate that they continue work.

If you have setup failures, please compare the configuration in the **passing CI
builds** with your configuration. That is almost certainly the source of the error.

### Tested on
__MacOS__
- Intel chip, MacOS Ventura 13.6
- Intel chip, MacOS Ventura 13.5.2
- Intel chip, MacOS Ventura 13.0

> - Intel chip, MacOS Monterey 12.6.7

Reviewer: I am really reluctant to put versions into the README because it is not clear who is taking on the responsibility of keeping this updated. Whose laptops are these? How often do they build the app? When is the last time anybody with that laptop built this app? "Tested on" doesn't mean anything unless we are able to actually fill in those results.

Pre-requisites
---
- The version of xcode used by the CI.
    - to install a particular version, use [xcode-select](https://www.unix.com/man-page/OSX/1/xcode-select/)
    - or this [supposedly easier to use repo](https://github.com/xcpretty/xcode-install)
    - **NOTE**: the basic xcode install on Catalina was messed up for me due to a prior installation of command line tools. [These workarounds helped](https://github.com/nodejs/node-gyp/blob/master/macOS_Catalina.md).
- git
- Java 17. Tested with [OpenJDK 17 (Temurin) using AdoptOpenJDK](https://adoptium.net). 
- if you are not on the most recent version of OSX: `homebrew`
    - this allows us to install the current version of cocoapods without
      running into ruby incompatibilities - e.g.
      https://github.com/CocoaPods/CocoaPods/issues/11763

> __1. Export statements__

Reviewer: This is inconsistent with the rest of our build process. We should not expect people to copy-and-paste a lot of commands, everything should be in the setup/active scripts so that it can be tested by the CI. This was originally not in the setup since the SDK would be installed while installing android studio. But now that we install the SDK manually, we should include this among the other environment variables that we define.

```
export ANDROID_SDK_ROOT="/Users/<user_name>/Library/Android/sdk"
```
```
export ANDROID_HOME="/Users/<user_name>/Library/Android/sdk"
```
aka the path where you want the SDK to be installed. 

To setup JAVA_HOME (after installing the latest JDK ), run this command:
```
/usr/libexec/java_home
```
Find the location of the Java installation (Default will look something like this:)
```
/Library/Java/JavaVirtualMachines/...
```
and then export the package as:
```
export JAVA_HOME="<path you derived from the above command>"
```

> 

Author: Will remove it

are available in the [e-mission-server README](https://github.com/e-mission/e-mission-server/blob/master/README.md).

The dynamic config (see https://github.com/e-mission/nrel-openpath-deploy-configs) controls the server endpoint that the phone app will connect to. If you are running the app in an emulator on the same machine as your local server (i.e. they share a `localhost`), you can use one of the `dev-emulator-*` configs (these configs have no `server` specified so `localhost` is assumed).

> If you wish to connect to a different server, create your own config file according to https://github.com/e-mission/nrel-openpath-deploy-configs and specify the `server` field accordingly.

Reviewer: I think we should also have instructions on how to change the config beyond the server. Maybe add a link to:
https://github.com/e-mission/nrel-openpath-deploy-configs/#testing-configs

The dynamic config (see https://github.com/e-mission/nrel-openpath-deploy-configs) controls the server endpoint that the phone app will connect to. If you are running the app in an emulator on the same machine as your local server (i.e. they share a `localhost`), you can use one of the `dev-emulator-*` configs (these configs have no `server` specified so `localhost` is assumed).

> If you wish to connect to a different server, create your own config file according to https://github.com/e-mission/nrel-openpath-deploy-configs and specify the `server` field accordingly. [deploy-configs](https://github.com/e-mission/nrel-openpath-deploy-configs/#testing-configs) has more information on this. 

Reviewer: ```suggestion
If you wish to connect to a different server, create your own config file according to https://github.com/e-mission/nrel-openpath-deploy-configs and specify the `server` field accordingly. The [deploy-configs](https://github.com/e-mission/nrel-openpath-deploy-configs/#testing-configs) repo has more information on this. 
```

are available in the [e-mission-server README](https://github.com/e-mission/e-mission-server/blob/master/README.md).

In order to make end-to-end testing even easier, we have moved from phone app to a dynamic config setting. We use the dynamic [config](configs) where each file upholds a function. Refer [Doc](https://github.com/e-mission/nrel-openpath-deploy-configs)

> If you have the [e-mission-server](https://github.com/e-mission/e-mission-server) running at a specific URL or IP address, and they want to connect to it, you would have to specify that URL or IP in the server field of their dynamic config file.

Reviewer: > We use the dynamic [config](configs)

is a broken link

> where each file upholds a function

is just incorrect

> If you have the [e-mission-server](https://github.com/e-mission/e-mission-server) running at a specific URL or IP address, and they want to connect to it, you would have to specify that URL or IP in the server field of their dynamic config file.

This is very confused around the actors ("they" and "you") and needs more clarification around when and how to override. I can't see anybody reading this and knowing what to do.

@JGreenlee or @Abby-Wheelis can you please just give @niccolopaganini the text to use? This change has already taken ~ 3 weeks and the steps outlined here are broken until it is merged.

One advantage of using `skip` authentication in development mode is that any user email can be entered without a password. Developers can use one of the emails that they loaded test data for in step (3) above. So if the test data loaded was with `-u shankari@eecs.berkeley.edu`, then the login email for the phone app would also be `shankari@eecs.berkeley.edu`.

## Updating the e-mission-\* plugins or adding new plugins

> 

Reviewer: Extraneous change

[![osx-build-ios](https://github.com/e-mission/e-mission-phone/actions/workflows/ios-build.yml/badge.svg)](https://github.com/e-mission/e-mission-phone/actions/workflows/ios-build.yml)
[![osx-build-android](https://github.com/e-mission/e-mission-phone/actions/workflows/android-build.yml/badge.svg)](https://github.com/e-mission/e-mission-phone/actions/workflows/android-build.yml)

Pre-requisites
---
- The version of xcode used by the CI.
    - to install a particular version, use [xcode-select](https://www.unix.com/man-page/OSX/1/xcode-select/)
    - or this [supposedly easier to use repo](https://github.com/xcpretty/xcode-install)
- git
- Java 17. Tested with [OpenJDK 17 (Temurin) using AdoptOpenJDK](https://adoptium.net). 
- Always use [homebrew](https://brew.sh) in addition to CLI
    - this allows us to install the current version of cocoapods without
      running into ruby incompatibilities - e.g.
      https://github.com/CocoaPods/CocoaPods/issues/11763

> :triangular_flag_on_post: __Important__

Reviewer: why has this been moved in the structure?

Most of the recent issues encountered have been due to incompatible setup. We
have now:
- locked down the dependencies,
- created setup and teardown scripts to setup self-contained environments with
  those dependencies, and
- CI enabled to validate that they continue work.

If you have setup failures, please compare the configuration in the **passing CI
builds** with your configuration. That is almost certainly the source of the error.

> __Export statements__

Reviewer: I don't think this warrants a bolded heading, especially when it is going to be removed anyway.

```
export ANDROID_SDK_ROOT="/Users/<user_name>/Library/Android/sdk"

> ```

Reviewer: We want to minimize manual copy-paste so these should go into the setup scripts instead.
We don't make people export anything else, so why this?
Please see my earlier comment this variable and JAVA_HOME.

This change has been pending for ~ 3 weeks, so I am fine with cleaning that up later.

are available in the [e-mission-server README](https://github.com/e-mission/e-mission-server/blob/master/README.md).

> In order to make end-to-end testing easy, if the local server is started on a HTTP (versus HTTPS port), it is in development mode.  By default, the phone app connects to the local server (localhost on iOS, [10.0.2.2 on android](https://stackoverflow.com/questions/5806220/how-to-connect-to-my-http-localhost-web-server-from-android-emulator-in-eclips)) with the `prompted-auth` authentication method. To connect to a different server, or to use a different authentication method, you need to modify the [nrel-commute.nrel-op.json](https://github.com/e-mission/nrel-openpath-deploy-configs/blob/482971d9715e8d52862e689658f9b4f2437e6401/configs/nrel-commute.nrel-op.json) file's dynamic config. 

Reviewer: This

> To connect to a different server, or to use a different authentication method, you need to modify the [nrel-commute.nrel-op.json](https://github.com/e-mission/nrel-openpath-deploy-configs/blob/482971d9715e8d52862e689658f9b4f2437e6401/configs/nrel-commute.nrel-op.json) file's dynamic config.

is just flat out wrong.

We are not going to modify the e-mission server nrel commute dynamic config to support random end-to-end testing.

```
"connectUrl": "https://nrel-commute-openpath.nrel.gov/api/"
```

> More details on configuring authentication [can be found in the docs](https://github.com/e-mission/e-mission-docs/blob/master/docs/install/configuring_authentication.md).

Reviewer: large parts of this document are obsolete.The phone auth is now based completely on the dynamic config.

[![osx-build-ios](https://github.com/e-mission/e-mission-phone/actions/workflows/ios-build.yml/badge.svg)](https://github.com/e-mission/e-mission-phone/actions/workflows/ios-build.yml)
[![osx-build-android](https://github.com/e-mission/e-mission-phone/actions/workflows/android-build.yml/badge.svg)](https://github.com/e-mission/e-mission-phone/actions/workflows/android-build.yml)

__Important__

Most of the recent issues encountered have been due to incompatible setup. We
have now:
- locked down the dependencies,
- created setup and teardown scripts to setup self-contained environments with
  those dependencies, and
- CI enabled to validate that they continue work.

If you have setup failures, please compare the configuration in the **passing CI
builds** with your configuration. That is almost certainly the source of the error.

Pre-requisites
---
- The version of xcode used by the CI.
    - to install a particular version, use [xcode-select](https://www.unix.com/man-page/OSX/1/xcode-select/)
    - or this [supposedly easier to use repo](https://github.com/xcpretty/xcode-install)
- git
- Java 17. Tested with [OpenJDK 17 (Temurin) using AdoptOpenJDK](https://adoptium.net). 

> - Always use [homebrew](https://brew.sh) in addition to CLI

Reviewer: "in addition to CLI"? What does that mean?

Reviewer: This has not been addressed

Run the setup script for the platform you want to build

```
bash setup/setup_android_native.sh
```
AND/OR
```
bash setup/setup_ios_native.sh
```

### Activation (after install, and in every new shell)

```
source setup/activate_native.sh
```

<details><summary> Expected Output </summary>

``` 
Activating nvm
Using version <X version number>
Now using node <X version number> (npm <Y version>)
npm version = <Y version>
Adding cocoapods to the path
Verifying /Users/<username>/Library/Android/sk or /Users/<username>/Library/Android/sdk is set
Activating sdkman, and by default, gradle
Ensuring that we use the most recent version of the command line tools
Configuring the repo for building native code
Copied config.cordovabuild.xml -> config.xml and package.cordovabuild.json -> package.json
```

</details>

> ### Activation (after install, and in every new shell)

Reviewer: this is not actually activation, it is a one time step

```suggestion
### Enable HTTP support on android by editing `config.xml`
```

More "versions" are available in [`package.cordovabuild.json`](package.cordovabuild.json)
```
    <edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application">
        <application android:usesCleartextTraffic="true"/>
    </edit-config>
npm run <type_of_build>
```

### Run in the emulator
For instance: (build-dev-android)
```
npm run build-dev-android
```
<details><summary> Your expected output should look something like this</summary>

```
$ npx cordova emulate ios
AND/OR
$ npx cordova emulate android
BUILD SUCCESSFUL in 2m 48s
52 actionable tasks: 52 executed
Built the following apk(s):
/Users/<Username>/e-mission-phone/platforms/android/app/build/outputs/apk/debug/app-debug.apk
```

Creating logos
---
</details> 

<br>

## Creating logos

If you are building your own version of the app, you must have your own logo to
avoid app store conficts. Updating the logo is very simple using the [`ionic
cordova resources`](https://ionicframework.com/docs/v3/cli/cordova/resources/)
command.

**Note**: You may have to install the [`cordova-res` package](https://github.com/ionic-team/cordova-res) for the command to work

Troubleshooting
---

> - Make sure to use `npx ionic` and `npx cordova`. This is

Reviewer: why has this not been restored to its original place?

## Beta-testing debugging
If users run into problems, they have the ability to email logs to the
maintainer. These logs are in the form of an sqlite3 database, so they have to
be opened using `sqlite3`. Alternatively, you can export it to a csv with
dates using the `bin/csv_export_add_date.py` script.

```
<download the log file>
mv ~/Downloads/loggerDB /tmp/logger.<issue>
pwd
.../e-mission-phone
python bin/csv_export_add_date.py /tmp/loggerDB.<issue>
less /tmp/loggerDB.<issue>.withdate.log
```

## Contributing

1. Add the main repo as upstream
```
git remote add upstream https://github.com/e-mission/e-mission-phone
```
2. Create a new branch (IMPORTANT). Please do not submit pull requests from master
```
git checkout -b <branch>
```
3. Make changes to the branch and commit them
```
git commit
```
4. Push the changes to your local fork
```
git push origin <branch>
```
5. Generate a pull request from the UI

<br>

__\*__Address my review comments__\*__

Once I merge the pull request :smiley: :tada:, pull the changes to your fork and delete the branch
```
git checkout master
```
```
git pull upstream master
```
```
git push origin master
```
```
git branch -d <branch>
```

---

> ### Troubleshooting

Reviewer: This is troubleshooting for the native build - why has it been moved to the bottom?

Reviewer: This has not been addressed

__1. Xcode command line tools__
```
Warning: No developer tools installed.
You should install the Command Line Tools.
```
```
xcode-select --install
```

> __2. Creating Logos__

Reviewer: why is "Creating logos" here? The rest of the text has nothing do to with logos

    ```

    </details>

__2. Installing (one time only)__

- Run the setup script for the platform you want to build

```
bash setup/setup_android_native.sh
AND/OR
bash setup/setup_ios_native.sh
```

**(optional)** Configure by changing the files in `www/json`.
Defaults are in `www/json/*.sample`

```
ls www/json/*.sample
cp www/json/startupConfig.json.sample www/json/startupConfig.json
cp ..... www/json/connectionConfig.json
```

If connecting to a development server over http, make sure to turn on http support on android

```
    <edit-config file="app/src/main/AndroidManifest.xml" mode="merge" target="/manifest/application">
        <application android:usesCleartextTraffic="true"/>
    </edit-config>
```
__3. Run this in every new shell__ 

> - __Activation__

Reviewer: I am not sure why you changed the formatting around this. The previous text did indicate that you needed to activate in every new shell.

```
source setup/activate_native.sh
```
<details><summary> Expected Output </summary>

``` 
Activating nvm
Using version 19.5.0
Now using node v19.5.0 (npm v9.3.1)
npm version = 9.3.1

> Adding cocoapods to the path

Reviewer: This has not been addressed

Verifying /Users/<username>/Library/Android/sk or /Users/<username>/Library/Android/sdk is set
Activating sdkman, and by default, gradle
Ensuring that we use the most recent version of the command line tools
Configuring the repo for building native code
Copied config.cordovabuild.xml -> config.xml and package.cordovabuild.json -> package.json
```

</details>

> - __Pick a type of build and execute the following:__

Reviewer: And you don't need to build as part of activation and you don't need to activate before every build.

More "versions" are available in [`package.cordovabuild.json`](https://github.com/e-mission/e-mission-phone/blob/fce117ff859abd995613bd405dbc7d27c703b09b/package.cordovabuild.json)
```
npm run <type_of_build>
```

For instance: (build-dev-android)
```
npm run build-dev-android
```
Your output should look something like this:

> ![Build Successful Message screenshot](Build_ss.png)

Reviewer: As I have said before, image-based screenshots are not the best way to represent text-based outputs. Please replace with text, like the other expected outputs.

## 4. End to End Testing

A lot of the visualizations that we display in the phone client come from the server. In order to do end to end testing, we need to run a local server and connect to it. Instructions for:

1. installing a local server,
2. running it, 
3. loading it with test data, and
4. running analysis on it

are available in the [e-mission-server README](https://github.com/e-mission/e-mission-server/blob/master/README.md).

> In order to make end to end testing easy, if the local server is started on a HTTP (versus HTTPS port), it is in development mode.  By default, the phone app connects to the local server (localhost on iOS, [10.0.2.2 on android](https://stackoverflow.com/questions/5806220/how-to-connect-to-my-http-localhost-web-server-from-android-emulator-in-eclips)) with the `prompted-auth` authentication method. To connect to a different server, or to use a different authentication method, you need to create a `www/json/connectionConfig.json` file. More details on configuring authentication [can be found in the docs](https://github.com/e-mission/e-mission-docs/blob/master/docs/install/configuring_authentication.md).

Reviewer: a fair amount of this is obsolete given the dynamic config. Did you check with @JGreenlee?

Reviewer 2: Is `connectionConfig.json` used at all anymore?

Reviewer: After the `master_for_platform` -> `master` change, I don't think so.

```
$ grep -r connectionConfig www/js
```

doesn't find any usages.

We should test removing it completely (in addition to fixing the README)

> One advantage of using `skip` authentication in development mode is that any user email can be entered without a password. Developers can use one of the emails that they loaded test data for in step (3) above. So if the test data loaded was with `-u shankari@eecs.berkeley.edu`, then the login email for the phone app would also be `shankari@eecs.berkeley.edu`.

Reviewer: This is also now different given the one-step opcode. Have you actually successfully run end-to-end with these instructions?

## 5. Beta-testing debugging
If users run into problems, they have the ability to email logs to the
maintainer. These logs are in the form of an sqlite3 database, so they have to
be opened using `sqlite3`. Alternatively, you can export it to a csv with
dates using the `bin/csv_export_add_date.py` script.

```
<download the log file>
mv ~/Downloads/loggerDB /tmp/logger.<issue>
pwd
.../e-mission-phone
python bin/csv_export_add_date.py /tmp/loggerDB.<issue>
less /tmp/loggerDB.<issue>.withdate.log
```

## 6. Contributing

1. Add the main repo as upstream
```
2. git remote add upstream <link/SSH key...>
```
3. Create a new branch (IMPORTANT). Please do not submit pull requests from master
```
4. git checkout -b <branch>
```
5. Make changes to the branch and commit them
```
6. git commit
```
 7. Push the changes to your local fork
```
8. git push origin <branch>
```
9. Generate a pull request from the UI

__\*__Address my review comments__\*__

Once I merge the pull request, pull the changes to your fork and delete the branch
```
git checkout master
```
```
git pull upstream master
```
```
git push origin master
```
```
git branch -d <branch>
```

> 

Reviewer: Why are these in separate code blocks?

---
### Troubleshooting
1. Xcode command line tools
```
Warning: No developer tools installed.
You should install the Command Line Tools.
```
```
xcode-select --install
```

> 2. Creating Logos

Reviewer: why is this here?

Author: > why is this here?

From my experience upgrading from Monterey to Ventura (on two devices), I noticed that the update removes the CLI tools. Given that this has nothing to do with e-mission-phone and more to do with Mac devices, I can remove it in the next PR. I only added this because ease I remember you mentioning something regarding Mac specific and thought it would be useful.

> ### Run in the emulator

Reviewer 2: ```suggestion
### Building the app
```

```
npm run <ios>
```
AND/OR
```
npm run <android>

> ```

Reviewer 2: It is not clear what `<ios>` and `<android>` are.

I think we should just give examples of the most common build scripts that we think will be used and list them out here.

`npm run build` (to build for both Android and iOS platforms)
`npm run build-prod-android` (to build for Android platform only)
`npm run build-prod-ios` (to build for iOS platform only)

> for builds, refer [`package.cordovabuild.json`](package.cordovabuild.json)

Reviewer 2: ```suggestion
For other options of build scripts, refer to [`package.cordovabuild.json`](package.cordovabuild.json)
```
